Lombiq.BaseTheme/Middlewares/RemoveBootstrapMiddleware.cs

```diff
@@ -0,0 +1,78 @@
+﻿using Lombiq.BaseTheme.Constants;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using OrchardCore.DisplayManagement.Manifest;
+using OrchardCore.Environment.Extensions;
+using OrchardCore.ResourceManagement;
+using OrchardCore.Themes.Services;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Lombiq.BaseTheme.Middlewares;
+
+/// <summary>
+/// Removes the built-in Bootstrap resource if the currently selected theme uses Lombiq.BaseTheme as its base theme.
+/// Themes derived from Lombiq.BaseTheme use Bootstrap from NPM as it's compiled into their site stylesheet. So this
+/// duplicate resource is not needed and can cause problems if not removed. This situation can arise when a module
+/// (such as Lombiq.DataTables) depends on Bootstrap and doesn't explicitly depend on Lombiq.BaseTheme so the built-in
+/// resource would injected if this middleware didn't remove it.
+/// </summary>
+public class RemoveBootstrapMiddleware
+{
+    private static readonly object _lock = new();
+    private static readonly string[] _resourceTypes = { "stylesheet", "script" };
+
+    private readonly RequestDelegate _next;
+
+    private static bool _done;
+
+    public RemoveBootstrapMiddleware(RequestDelegate next) => _next = next;
+
+    public Task InvokeAsync(
+        HttpContext context,
+        IOptions<ResourceManagementOptions> resourceManagementOptions,
+        ISiteThemeService siteThemeService) =>
+        _done
+            ? _next(context)
+            : InvokeInnerAsync(context, resourceManagementOptions.Value, siteThemeService);
+
+    private async Task InvokeInnerAsync(
+        HttpContext context,
+        ResourceManagementOptions resourceManagementOptions,
+        ISiteThemeService siteThemeService)
+    {
+        if (IsCurrentTheme(await siteThemeService.GetSiteThemeAsync()))
+        {
+            var resourcesToClear = resourceManagementOptions
+                .ResourceManifests
+                .Where(manifest => !manifest.GetResources("$" + nameof(FeatureIds.Area)).ContainsKey(FeatureIds.Area))
+                .SelectMany(manifest => _resourceTypes
+                    .SelectWhere(manifest.GetResources)
+                    .SelectWhere(resources => resources.GetMaybe("bootstrap"), resource => resource?.Any() == true))
+                .ToList();
+
+            // This only happens once per process instance, so locking would be rare.
+            if (resourcesToClear.Any()) ClearResources(resourcesToClear);
```

Comment:
Do I understand correctly that this removes the Bootstrap resource globally? So, even the admin will use the BS version provided by BaseTheme.

Reply:
The admin doesn't use the Bootstrap resource. The style is baked into TheAdmin.css. So this has no effect on the admin.
![image](https://user-images.githubusercontent.com/4007293/208293874-748ff691-6b71-401b-9048-88bf23d668f4.png)

Reply:
Admin feature can though, and Workflows [also does](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Workflows/Views/Workflow/Details.cshtml#L99). While apparently this isn't common, it's not strange to do this, so our override seems risky.

I feel that here we're really trying to make something work, and work well and safe, but we shouldn't really override resources in such ways the first place. I can't offer a better solution though. Perhaps something our modules like DataTables should be somehow made aware? Or Orchard should work in a way that the theme can override resources out of the box?

Reply:
> Admin feature can though, and Workflows [also does](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Workflows/Views/Workflow/Details.cshtml#L99).

That makes no sense to me, why is it using Bootstrap 4?! I can remove Bootstrap 5 only, instead of all versions. Lombiq.BaseTheme's resource manifest adds the NPM version of the BS5 script anyway and I'm still confident that the BS5 CSS should not be available to TheAdmin since it's baked into their own theme. So that way we have full coverage.

> I feel that here we're really trying to make something work, and work well and safe, but we shouldn't really override resources in such ways the first place.

Well yeah, I think this is not a supported use-case, but if it was easy and supported we wouldn't need this base theme. They are not the bosses of us anyway. 😎 

> Perhaps something our modules like DataTables should be somehow made aware?

We _could_ make our own modules aware of Lombiq.BaseTheme, but not all modules in the whole OC ecosystem. So we shouldn't do it for our own modules either. That way our own modules can reveal if there is some incompatibility that may be a problem with third party modules too.

> Or Orchard should work in a way that the theme can override resources out of the box?

In principle I agree, but it seems to me that it's intentionally not possible. So any proposal to make this possible will be absolutely shut down by the OC product owners. It's not even worth raising the issue imo.

Reply:
This whole setup we have is really fragile, and it's not much better with a BS version constraint. Now the theme being enabled can break the Workflows admin triggered by an OC upgrade. This is something that's very hard to notice. We continue to patch an approach that I'm really unsure about.

Please take a step back and think this story through from the very beginning: Is removing a resource the only way you can see this problem being tackled? We can also get on a call if you'd like to brainstorm.

Reply:
> Now the theme being enabled can break the Workflows admin triggered by an OC upgrade. 

An update wouldn't break it. That's why I wrote "Lombiq.BaseTheme's resource manifest adds the NPM version of the BS5 script". If they update that page to use version 5 then it will use the resource from Lombiq.BaseTheme instead of the one removed from stock. So we remove one script but also provide the replacement. I think the current arrangement is robust and not fragile. Message me and let's have a call if you are still not convinced.

Reply:
E.g. if Orchard upgrades to BS 5.1.4, Workflows starts to reference that instead of 4, then the middleware here will make it load 5.1.3 instead. My point is, that the risk here is that this is code that can cause the app to break in hard-to-notice ways in completely different places when you don't expect it. It causes global side effects.

That's why I ask you to think if you can come up with a fundamentally different solution that doesn't have this.

Reply:
Sounds like a rare edge case that can only cause problems if OC upgrades to a newer version of BS than what we have _and_ also uses some script feature in the admin that's only available in that newer version. This doesn't strike me as a realistic concern. More importantly, I don't consider this a side effect, because I think maintaining the NPM BS in Lombiq.BaseTheme is part of maintaining this repo so we should always be either on par or ahead of the BS5 in OrchardCore.Resources.

I have rewritten the removal logic to keep the v5.x BS script that has the higher version number instead of indiscriminately keeping ours. This way if we get behind the stock, then the stock is retained and ours is removed. I have been around this topic for months and I have already pondered if there can be any other alternatives. Short of removing the NPM Bootstrap (which means giving up on the ability to override their SASS variables or use their SASS mixins) I don't think it's possible to do anything fundamentally different that's also more robust. But then again, I still don't think this solution is realistically fragile at all and I think your concerns are exaggerated.

Reply:
Let's see.